treeopt/visualize.py

```diff
@@ -0,0 +1,370 @@
+                        :, 0
+                    ].reshape(-1),
+                    np.atleast_2d(self.opti_data.x[: self.opti_data.numDOE])[
+                        :, 1
+                    ].reshape(-1),
+                    marker="x",
+                    color="r",
+                    label="Points-DOE",
+                )
+            )
+            images.append(
+                axes[i].scatter(
+                    np.atleast_2d(self.opti_data.x[self.opti_data.numDOE :])[
+                        :, 0
+                    ].reshape(-1),
+                    np.atleast_2d(self.opti_data.x[self.opti_data.numDOE :])[
+                        :, 1
+                    ].reshape(-1),
+                    marker="x",
+                    color="m",
+                    label="Expanded Points",
+                )
+            )
+            images.append(
+                axes[i].scatter(
+                    np.atleast_2d(self.opti_data.current_best_point[0]),
+                    self.opti_data.current_best_point[1],
+                    marker="x",
+                    color="white",
+                )
+            )
+
+        images.append(axes[0].set_title("Benchmarking function"))
+        images.append(axes[1].set_title("Calculated metamodell"))
+
+        fig.subplots_adjust(bottom=0.225, top=0.85)
+
+        axes = np.hstack([axes, fig.add_axes([0.125, 0.075, 0.775, 0.075])])
+
+        images.append(
+            mpl.colorbar.ColorbarBase(
+                axes[-1],
+                cmap=mpl.cm.viridis,
+                norm=norm,
+                orientation="horizontal",
+            )
+        )
+
+        if hasattr(self.opti_data, "name"):
+            fig.suptitle(self.opti_data.name)
+
+        lines = []
+        labels = []
+
+        axLine, axLabel = axes[0].get_legend_handles_labels()
+        lines.extend(axLine)
+        labels.extend(axLabel)
+
+        fig.legend(
+            lines,
+            labels,
+            loc="lower center",
+            bbox_to_anchor=(0.5, -0.2),
+            ncol=2,
+        )
+
+        plt.show()
+
+    def plot_metamodell_1vars(self):
+        """
+        Plots a metamodell with two one input variable and one output variable
+        :return: Nothing
+        :rtype: None
+        """
+
+        x_step = np.linspace(
+            self.opti_data.limits[0, 0],
+            self.opti_data.limits[0, 1],
+            self.num_ele,
+        )
+
+        fun1Vars = self.opti_data.sm.predict_values(x_step)
+
+        fig, axes = plt.subplots(nrows=1, ncols=1)
+
+        images = []
+
+        images.append(plt.plot(x_step, fun1Vars, "-", c="blue"))
+        images.append(plt.plot(self.opti_data.x, self.opti_data.y, "x", c="r"))
+        images.append(plt.grid())
+
+        if hasattr(self.opti_data, "name"):
+            fig.suptitle(self.opti_data.name)
+
+        plt.show()
+
+    def plot_metamodell_2vars(self):
+        """
```

Comment:
See above. Also *(I admit, I didn't check the contents of the functions in detail)*, I wonder if some of the code can be shared between the different functions. It seems a bit repetitive. What do you think?

Reply:
I admit, that the visualization methods are probably the least comprehensible. But to visualize simple problems they work. We should discuss to which extent visualization methods are needed treeopt.

Reply:
I think that is fine! Visualization code is almost always hard to read, and the functionality is useful. If there is a bug here, it will not influence the optimization. Let's keep it!